detour_test.go

```diff
@@ -71,14 +76,52 @@ func TestBlockedImmediately(t *testing.T) {
 	}
 }
 
+func TestReadFailedImmediately(t *testing.T) {
+	defer RemoveFromWl("127.0.0.1")
+	defer stopMockServers()
+	proxiedURL, _ := newMockServer(detourMsg)
+	firstReadTimeoutToDetour = 50 * time.Millisecond
+	mockURL, _ := newMockServer(detourMsg)
+
+	client := newDirectFailingClient(proxiedURL, 1*time.Hour, 0)
+	u, _ := url.Parse(mockURL)
+	resp, err := client.Get(mockURL)
+	if assert.NoError(t, err, "should have no error if reading fails immediately") {
+		defer resp.Body.Close()
+		assert.True(t, wlTemporarily(u.Host), "should be added to whitelist if reading fails immediately")
+		assertContent(t, resp, detourMsg, "should detour if reading fails immediately")
+	}
+}
+
+func TestReadFailedEventually(t *testing.T) {
+	defer RemoveFromWl("127.0.0.1")
+	defer stopMockServers()
+	longMessage := make([]byte, 10000)
+	rand.Read(longMessage)
+	proxiedURL, _ := newMockServer(string(longMessage))
+	firstReadTimeoutToDetour = 50 * time.Millisecond
+	mockURL, _ := newMockServer(directMsg)
+
+	client := newDirectFailingClient(proxiedURL, 1*time.Hour, 1)
+	u, _ := url.Parse(mockURL)
+	resp, err := client.Get(mockURL)
+	if assert.NoError(t, err, "should get response if reading fails after first read") {
+		defer resp.Body.Close()
+		_, copyErr := io.Copy(ioutil.Discard, resp.Body)
+		if assert.Error(t, copyErr, "reading should have failed eventually") {
+			assert.True(t, wlTemporarily(u.Host), "should be added to whitelist if reading fails after first read")
+		}
+	}
+}
+
 func TestRemoveFromWhitelist(t *testing.T) {
 	defer RemoveFromWl("127.0.0.1")
 	defer stopMockServers()
 	proxiedURL, proxy := newMockServer(detourMsg)
 	proxy.Timeout(200*time.Millisecond, detourMsg)
 	firstReadTimeoutToDetour = 50 * time.Millisecond
 	mockURL, _ := newMockServer(directMsg)
-	client := newClient(proxiedURL, 100*time.Millisecond)
+	client := newDetourFailingClient(proxiedURL, 1*time.Hour, 0)
```

Comment:
I changed the failure mode here because in the original test, we were depending on the client closing the connection in order to consider it failed.